modules/core/src/main/scala/org/tessellation/infrastructure/trust/storage/TrustStorage.scala

```diff
@@ -4,49 +4,120 @@ import cats.Monad
+          val t = store.trust
+          val updated = t.trust ++ trustUpdates.map {
+            case (k, v) =>
+              k -> t.trust.getOrElse(k, TrustInfo()).focus(_.predictedTrust).replace(v.some)
+          }
+          store.focus(_.trust.trust).replace(updated)
         }
 
-      def getTrust: F[Map[PeerId, TrustInfo]] = trust.get
+      def getTrust: F[TrustMap] = trustStoreRef.get.map(_.trust)
+
+      def getCurrentOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.currentOrdinalTrust)
+
+      def getNextOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.nextOrdinalTrust)
+
+      def getPublicTrust: F[PublicTrust] = getTrust.map(_.toPublicTrust)
 
-      def getPublicTrust: F[PublicTrust] = getTrust.map { trustValue =>
-        PublicTrust(trustValue.mapFilter(_.publicTrust))
+      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trustStoreRef.update { store =>
+        val t = store.trust
+        val updated = t.peerLabels.value + (peerId -> publicTrust)
+        store.focus(_.trust.peerLabels).replace(PublicTrustMap(updated))
       }
 
-      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trust.update { t =>
-        t.updatedWith(peerId) { trustInfo =>
-          trustInfo
-            .orElse(Some(TrustInfo()))
-            .map(_.copy(peerLabels = publicTrust.labels))
+      def updateNext(ordinal: SnapshotOrdinal): F[Option[SnapshotOrdinalPublicTrust]] =
+        trustStoreRef.modify { store =>
+          val next = store.nextOrdinalTrust
+
+          if (ordinal >= next.ordinal && next.trust == TrustMap.empty) {
+            val updated = next.focus(_.trust).replace(store.trust)
+            val ordinalPublicTrust = SnapshotOrdinalPublicTrust(
+              next.ordinal,
+              store.trust.toPublicTrust
+            )
+
+            (store.focus(_.nextOrdinalTrust).replace(updated), ordinalPublicTrust.some)
+          } else (store, none)
         }
+
+      def updateNext(
+        peerId: PeerId,
+        publicTrust: SnapshotOrdinalPublicTrust
+      ): F[Unit] = trustStoreRef.update { store =>
+        val next = store.nextOrdinalTrust
+
+        val updated =
+          if (next.ordinal === publicTrust.ordinal && !next.peerLabels.value.contains(peerId)) {
+            val update = next.peerLabels.add(peerId, publicTrust.labels)
+
+            next.focus(_.peerLabels).replace(update)
+          } else next
+
+        store.focus(_.nextOrdinalTrust).replace(updated)
       }
 
-      def updatePredictedTrust(trustUpdates: Map[PeerId, Double]): F[Unit] = trust.update { t =>
-        t ++ trustUpdates.map {
-          case (k, v) =>
-            k -> t.getOrElse(k, TrustInfo()).copy(predictedTrust = v.some)
+      def updateCurrent(ordinal: SnapshotOrdinal): F[Unit] =
+        trustStoreRef.update { store =>
+          if (ordinal >= store.nextOrdinalTrust.ordinal.plus(ordinalTrustUpdateDelay)) {
```

Comment:
as above (ordinal)

Reply:
The `SnapshotOrdinal`s are compared with `Order`.